### PR TITLE
[SPARK-18057][FOLLOW-UP][SS] Update Kafka client version from 0.10.0.1 to 2.0.0

### PR DESCRIPTION
**Pull request by srowen**
## What changes were proposed in this pull request?

Update to kafka 2.0.0 in streaming-kafka module, and remove override for Scala 2.12. It won't compile for 2.12 otherwise.

## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by tedyu** [comment]
Thanks for the follow-up.

lgtm

**Review by ijuma — APPROVED** [review]
LGTM

**Review by felixcheung — APPROVED** [review]

**Comment by srowen** [comment]
@ijuma heh yeah I just pushed this to a branch to see what would happen before signing off. Not quite there. I pushed fixes for a few of the obvious compile issues, but might need a Kafka expert to weigh in on how to modify the other calls. Probably an easy change but requires calling slightly different APIs.

**Comment by srowen** [comment]
BTW this only passes, I think, because I snipped out part of the log compaction test. I don't think it's valid to merge as is. Pinging @koeninger in case he's willing and able to weigh in on how to write this test correctly for the 2.x API. I'm just missing how you write (key, value) bytes to a Log.

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaRDDSuite.scala [hunk 1]
```diff
@@ -72,31 +72,37 @@ class KafkaRDDSuite extends SparkFunSuite with BeforeAndAfterAll {
...
+      Int.MaxValue,
+      new TopicPartition(topic, partition),
+      new ProducerStateManager(new TopicPartition(topic, partition), dir),
+      logDirFailureChannel
     )
     messages.foreach { case (k, v) =>
-      val msg = new ByteBufferMessageSet(
-        NoCompressionCodec,
-        new Message(v.getBytes, k.getBytes, Message.NoTimestamp, Message.CurrentMagicValue))
-      log.append(msg)
```
Comment (ijuma) [inline external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaRDDSuite.scala]: there is `Log.appendAsLeader` and `Log.appendAsFollower` depending on your goal here.

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaRDDSuite.scala [hunk 2]
```diff
@@ -72,31 +72,37 @@ class KafkaRDDSuite extends SparkFunSuite with BeforeAndAfterAll {
...
-    val logs = new Pool[TopicAndPartition, Log]()
+    val logs = new Pool[TopicPartition, Log]()
     val logDir = kafkaTestUtils.brokerLogDir
     val dir = new File(logDir, topic + "-" + partition)
     dir.mkdirs()
     val logProps = new ju.Properties()
     logProps.put(LogConfig.CleanupPolicyProp, LogConfig.Compact)
     logProps.put(LogConfig.MinCleanableDirtyRatioProp, java.lang.Float.valueOf(0.1f))
+    // TODO is this new Log declaration correct?
+    val logDirFailureChannel = new LogDirFailureChannel(0)
```
Comment (ijuma) [inline external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaRDDSuite.scala]: This should be `1` if we're assuming a single log directory.

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaRDDSuite.scala [hunk 3]
```diff
@@ -72,33 +72,39 @@ class KafkaRDDSuite extends SparkFunSuite with BeforeAndAfterAll {
...
+      new TopicPartition(topic, partition),
+      new ProducerStateManager(new TopicPartition(topic, partition), dir),
+      logDirFailureChannel
     )
     messages.foreach { case (k, v) =>
-      val msg = new ByteBufferMessageSet(
-        NoCompressionCodec,
-        new Message(v.getBytes, k.getBytes, Message.NoTimestamp, Message.CurrentMagicValue))
-      log.append(msg)
+      val records = new MemoryRecords()
```
Comment (ijuma) [inline external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaRDDSuite.scala]:
```java
public static MemoryRecords withRecords(CompressionType compressionType, SimpleRecord... records) {
```
Maybe you can use the above?

**Comment by ijuma** [comment]
@srowen I think you're pretty close, I added a few comments.

**Comment by zsxwing** [comment]
There are some configs need to tune to make the tests faster. See: https://github.com/apache/spark/pull/21488/files#diff-1aecc5a638048c27f97cbd9579cb53b5R311

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaTestUtils.scala [hunk 4]
```diff
@@ -109,7 +109,7 @@ private[kafka010] class KafkaTestUtils extends Logging {
       brokerConf = new KafkaConfig(brokerConfiguration, doLog = false)
       server = new KafkaServer(brokerConf)
       server.startup()
-      brokerPort = server.boundPort()
+      brokerPort = server.boundPort(brokerConf.interBrokerListenerName)
```
[the remark this answers is not in the document]
Comment (zsxwing) [reply external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaTestUtils.scala]: > @zsxwing comparing to your change... should this be "PLAINTEXT"?

I think both fine.

> default.api.timeout.ms -> 3000

You don't need to add `default.api.timeout.ms` to `brokerConfiguration`. It's a client configuration. I think you don't need to add this to any place. DStreams Kafka doesn't have tests requiring this config.

**Review by koeninger — APPROVED** [review]

**Comment by koeninger** [comment]
jenkins, retest this please

**Comment by srowen** [comment]
I believe the second-to-last 'pass' is the authoritative one. The only change difference between the two runs is an added comment in an XML file. I'm going to proceed as this might fix up a test, and makes Scala 2.12 work again for Spark 2.4

**Comment by srowen** [comment]
Hm, I might need your eyes on this again @koeninger @zsxwing or even @ijuma . I see this failure in the test, which looks like it's no longer happy starting ZK in some cases. Does that ring any bells? not sure it would.

```
[info] org.apache.spark.streaming.kafka010.DirectKafkaStreamSuite *** ABORTED *** (8 seconds, 741 milliseconds)
[info]   kafka.zookeeper.ZooKeeperClientTimeoutException: Timed out waiting for connection while in state: CONNECTING
[info]   at kafka.zookeeper.ZooKeeperClient$$anonfun$kafka$zookeeper$ZooKeeperClient$$waitUntilConnected$1.apply$mcV$sp(ZooKeeperClient.scala:230)
[info]   at kafka.zookeeper.ZooKeeperClient$$anonfun$kafka$zookeeper$ZooKeeperClient$$waitUntilConnected$1.apply(ZooKeeperClient.scala:226)
[info]   at kafka.zookeeper.ZooKeeperClient$$anonfun$kafka$zookeeper$ZooKeeperClient$$waitUntilConnected$1.apply(ZooKeeperClient.scala:226)
[info]   at kafka.utils.CoreUtils$.inLock(CoreUtils.scala:251)
[info]   at kafka.zookeeper.ZooKeeperClient.kafka$zookeeper$ZooKeeperClient$$waitUntilConnected(ZooKeeperClient.scala:226)
[info]   at kafka.zookeeper.ZooKeeperClient.<init>(ZooKeeperClient.scala:95)
[info]   at kafka.zk.KafkaZkClient$.apply(KafkaZkClient.scala:1580)
[info]   at kafka.server.KafkaServer.kafka$server$KafkaServer$$createZkClient$1(KafkaServer.scala:348)
[info]   at kafka.server.KafkaServer.initZkClient(KafkaServer.scala:372)
[info]   at kafka.server.KafkaServer.startup(KafkaServer.scala:202)
[info]   at org.apache.spark.streaming.kafka010.KafkaTestUtils$$anonfun$setupEmbeddedKafkaServer$2.apply(KafkaTestUtils.scala:111)
```

See https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94150/ for instance.

However I note that this test has also passed as well, at least in one of the Maven-based test jobs on Jenkins. Could be a little build-specific but nevertheless a problem.

File: external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaTestUtils.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (srowen) [reply external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaTestUtils.scala]: In particular @zsxwing , regarding the failure below, I wonder if my change on this line really wasn't quite equivalent. Is it possible it needs to be "PLAINTEXT" like in yours?

[the remark this answers is not in the document]
Comment (koeninger) [reply external/kafka-0-10/src/test/scala/org/apache/spark/streaming/kafka010/KafkaTestUtils.scala]: Isn't the test hanging on the line right before that change though?

**Comment by koeninger** [comment]
I don't see an obvious issue.  Looks like zookeeper.connection.timeout.ms isn't being set, so it's defaulting to 6 seconds... could try tweaking it upwards.

**Comment by ijuma** [comment]
Looks similar to https://issues.apache.org/jira/browse/KAFKA-7193, are we using the same ZK ip in the server and client?

**Comment by srowen** [comment]
I see that `zookeeper.connect` is set to `[zkhost]:[zkport]` after ZK is started, so looks like it's trying to do so. That much hasn't changed at least. I can increase the timeout; although I don't know why it takes a while, it could explain why it appears flaky. It's also consistent with KAFKA-7193 although I think we're on ZK 3.4.5 or so in Spark?

**Comment by srowen** [comment]
Let's continue at https://github.com/apache/spark/pull/21995

**Pull request closed without merging**